### PR TITLE
read target cluster from Space

### PR DESCRIPTION
**Pull request by MatousJobanek**
[KUBESAW-107](https://issues.redhat.com/browse/KUBESAW-107)

[KUBESAW-108](https://issues.redhat.com/browse/KUBESAW-108)

Take the target cluster from Space (either the home one or the first Space the user has access to)

changes in go.sum are result of running `go mod tidy`

### CODE REVIEW

File: pkg/signup/service/signup_service.go [hunk 1]
```diff
@@ -612,58 +614,51 @@ func (s *ServiceImpl) PhoneNumberAlreadyInUse(userID, username, phoneNumberOrHas
...
-	var createdSpaceFound bool
-	for _, sb := range sbs {
-		spaceName := sb.Labels[toolchainv1alpha1.SpaceBindingSpaceLabelKey]
-		if spaceName == "" { // space may not be initialized
-			// log error and continue so that the api behaves in a best effort manner
-			log.Errorf(nil, fmt.Errorf("spacebinding has no '%s' label", toolchainv1alpha1.SpaceBindingSpaceLabelKey), "unable to get space '%s'", spaceName)
-			continue
-		}
-		space, err := provider.GetSpace(spaceName)
+func GetDefaultUserTarget(provider ResourceProvider, spaceName, murName string) (string, string) {
```
Comment (alexeykazakov) [inline pkg/signup/service/signup_service.go:617]: Can you add a comment to the function stating that it returns `<target-member-cluster>, <default-namespace>`?

**Review by alexeykazakov — APPROVED** [review]

File: pkg/signup/service/signup_service.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/signup/service/signup_service.go:625]: fac5e9e

File: pkg/signup/service/signup_service.go [hunk 2]
```diff
@@ -612,58 +614,59 @@ func (s *ServiceImpl) PhoneNumberAlreadyInUse(userID, username, phoneNumberOrHas
...
+			for i, sb := range sbs {
+				spaceNames[i] = sb.Spec.Space
 			}
+			sort.Strings(spaceNames)
+			spaceName = spaceNames[0]
 		}
+	}
+	if spaceName == "" {
+		return "", ""
+	}
```
Comment (mfrancisc) [inline pkg/signup/service/signup_service.go:651]: alternatively you could return if `sbs` are empty and avoid double checking the `spaceName`
```suggestion
		if len(sbs) == 0 {
			return "", ""
		}
		spaceNames := make([]string, len(sbs))
		for i, sb := range sbs {
			spaceNames[i] = sb.Spec.Space
		}
		sort.Strings(spaceNames)
		spaceName = spaceNames[0]
	}
```

File: pkg/signup/service/signup_service_test.go [hunk 3]
```diff
@@ -1665,74 +1683,54 @@ func (s *TestSignupServiceSuite) TestGetDefaultUserNamespace() {
...
 		// then
 		assert.Equal(t, "dave-dev", defaultUserNamespace)
+		assert.Equal(s.T(), "member-123", targetCluster)
 	})
 }
 
-// TestGetDefaultUserNamespaceOnlyUnownedSpace tests that the default user namespace is returned even if the only accessible Space was not created by the user.
-// This is valuable because if the creator label on a Space is missing for whatever reason, APIs that depend on this field will still work.
+// TestGetDefaultUserNamespaceOnlyUnownedSpace tests that the default user namespace is returned even if the only accessible Space was not created as the home space.
+// This is valuable when user doesn't have default home space created, but has access to some shared spaces
```
Comment (mfrancisc) [inline pkg/signup/service/signup_service_test.go:1714]: should we add a similar test, but with multipe spacebindings, just to make sure we always return the first one ? 

something like
```
func (s *TestSignupServiceSuite) TestGetDefaultUserNamespaceOnlyMultipleUnownedSpace() {
	// given
	s.ServiceConfiguration(configuration.Namespace(), true, "", 5)
	// space created for userA
	space := s.newSpace("userA")
	err := s.FakeSpaceClient.Tracker.Add(space)
	require.NoError(s.T(), err)

	spacebinding := s.newSpaceBinding("userB", space.Name)
	err = s.FakeSpaceBindingClient.Tracker.Add(spacebinding)
	require.NoError(s.T(), err)

	// space created for userC
	spaceC := s.newSpace("userC")
	err = s.FakeSpaceClient.Tracker.Add(spaceC)
	require.NoError(s.T(), err)

	// space shared with userB
	spacebindingC := s.newSpaceBinding("userB", spaceC.Name)
	err = s.FakeSpaceBindingClient.Tracker.Add(spacebindingC)
	require.NoError(s.T(), err)

	// when
	targetCluster, defaultUserNamespace := service.GetDefaultUserTarget(s, "", "userB")

	// then
	assert.Equal(s.T(), "userA-dev", defaultUserNamespace)
	assert.Equal(s.T(), "member-123", targetCluster)

	s.T().Run("informer", func(t *testing.T) {
		// given
		inf := fake.NewFakeInformer()
		inf.GetSpaceFunc = func(_ string) (*toolchainv1alpha1.Space, error) {
			return space, nil
		}
		inf.ListSpaceBindingFunc = func(_ ...labels.Requirement) ([]toolchainv1alpha1.SpaceBinding, error) {
			return []toolchainv1alpha1.SpaceBinding{*spacebinding}, nil
		}

		// when
		targetCluster, defaultUserNamespace := service.GetDefaultUserTarget(inf, "", "userB")

		// then
		assert.Equal(t, "userA-dev", defaultUserNamespace)
		assert.Equal(s.T(), "member-123", targetCluster)
	})
}
```

**Review by mfrancisc — APPROVED** [review]
Looks great 👍 

I have only some minor comments!

File: pkg/signup/service/signup_service.go (hunk 2 above)
Comment (MatousJobanek) [reply pkg/signup/service/signup_service.go]: yeah, no idea why I didn't realize it before: [ca51a4d](https://github.com/codeready-toolchain/registration-service/pull/441/commits/ca51a4d5ef26fd19c1c3e90d5b47540028aec397)
Thanks

File: pkg/signup/service/signup_service_test.go (hunk 3 above)
Comment (mfrancisc) [reply pkg/signup/service/signup_service_test.go:1714]: makes sense indeed! Good point!

**Review by mfrancisc — APPROVED** [review]

**Pull request merged**